Blog posts/using aws ec2 and ecs to host hundreds of services (#218)
deliveroo/deliveroo.engineering: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md

secure, resizable compute capacity in the cloud. It is designed to make
web-scale cloud computing easier for developers. Amazon EC2’s simple web service
interface allows you to obtain and configure capacity with minimal friction. It
provides you with complete control of your computing resources and lets you run
on Amazon’s proven computing environment."

_Amazon's description of EC2_
</aside>

I would say this; an EC2 host is a server. More simply, it is a computer. It is
(in most cases) not an actual physical server in a rack, and Amazon abstracts
that detail away us, but I find I get my head around the concept easier by
thinking of them as physical machines anyway. The machines we generally use have
16 vCPUs and 64 GiB of Memory (RAM).

It comes preinstalled with the software required to make it a usable computer;
like an operating system (you can just assume Linux for now - others are
available though), so it can be booted up and can run processes - more on that
later…

## What do we use EC2 hosts for?

A few different uses, but the most common use is in an
[ECS Cluster](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/clusters.html),

> a grouping of EC2 machines used as a home for ECS tasks - these are the

Reviewer: ```suggestion
a grouping of EC2 machines used as a home for ECS Tasks - these are the
```

of tasks - if the number of running tasks is less than this, ECS will start more
containers to reach the desired number, if there are more than desired, ECS will
safely terminate running containers to reach the desired number.

## Where does ECS start running this one container I’ve asked for?

This takes us back to our cluster of EC2 machines. ECS will find an EC2 machine
in the cluster that has enough spare capacity on it to hold and run your task
(i.e. it has enough spare reserved CPU and memory - which is specified in the
config file). There are some other rules in place regarding which Availability
Zone your task is running in (we don’t want all your eggs in one basket), but
for the most part, we leave it to ECS to decide.

## What happens if the cluster is full?

We are constantly monitoring the ECS cluster, and autoscale EC2 instances based
on how much spare capacity there is. If there’s not enough spare capacity to
immediately run another 40 large docker containers, we bump up the desired count
of EC2 instances in the cluster, and EC2 spins up new machines (the number of
machines we start up depends on how much below approximately 40 large container

> capacity we are). New EC2 instances can take up to 10 minutes before they’re

Reviewer: ```suggestion
capacity we are). New EC2 instances can take a few minutes before they’re
```

I think our current timing is about two and a half minutes.

ready to be used by ECS, so we need to have quite a large buffer to deal with

> dinnertime spikes in demand.

Reviewer: ```suggestion
unexpected spikes in demand.
```

## How do we change or upgrade the machine image used?

Circling back to the software that is preinstalled on these EC2 servers. When

> booted up, an Amazon Machine Image (AMI) is used, which installs the basic

Reviewer: The tools are baked into the AMI. On boot, the only required tasks are to register the speicifc hosts into the cluster, download some config and start running the Docker containers.

> tools you need to make the machine usable. Amazon provides a Linux AMI which we

Reviewer: You can mention that we use packer and ansible to create our own Amazon Linux derived images.

have built upon in the past to create our own machine image. We use this to run
scripts which give all our running ECS tasks access to things that all
Deliveroo’s services will need, such as software (like the

> [Datadog agent](https://docs.datadoghq.com/agent/) which sends metrics to

Reviewer: Do you want to also mention other agents that we run? e.g. AWS Inspector, Cloudstrike Falcon and a few Deliveroo specific agents (e.g. for healthchecks)

  done, and how we manage the process of updates and improvements to this
  system.

---

One of my goals of moving internally to the Production Engineering team was to
help demystify the concepts that are commonplace within our Platform teams. My
first internal blog post to do this was to share how we use EC2 (AWS Elastic
Compute Cloud) and ECS (AWS Elastic Container Service) to host the hundreds of
services that our Software Engineers build and improve every day.

## What is an EC2 host?

<aside>
"Amazon Elastic Compute Cloud (Amazon EC2) is a web service that provides
secure, resizable compute capacity in the cloud. It is designed to make
web-scale cloud computing easier for developers. Amazon EC2’s simple web service
interface allows you to obtain and configure capacity with minimal friction. It
provides you with complete control of your computing resources and lets you run
on Amazon’s proven computing environment."

_Amazon's description of EC2_
</aside>

I would say this; an EC2 host is a server. More simply, it is a computer. It is
(in most cases) not an actual physical server in a rack, and Amazon abstracts

> that detail away us, but I find I get my head around the concept easier by

Reviewer 2: ```suggestion
that detail away for us, but I find I get my head around the concept easier by
```

thinking of them as physical machines anyway. The machines we generally use have
16 vCPUs and 64 GiB of Memory (RAM).

It comes preinstalled with the software required to make it a usable computer;
like an operating system (you can just assume Linux for now - others are
available though), so it can be booted up and can run processes - more on that
later…

## What do we use EC2 hosts for?

A few different uses, but the most common use is in an
[ECS Cluster](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/clusters.html),
a grouping of EC2 machines used as a home for ECS Tasks - these are the
Dockerized containers of our applications that are running with a command
specified by the engineer in a config file.

> ## ECS? what’s that, and how is it related to EC2?

Reviewer 2: ```suggestion
## ECS? What’s that, and how is it related to EC2?
```

to be used by ECS, so we need to have a buffer to deal with unexpected spikes in
demand.

## How do we change or upgrade the machine image used?

Circling back to the software that is preinstalled on these EC2 servers. When
booted up, an Amazon Machine Image (AMI) is used, which has some basic tools
installed on it to make the machine usable. Amazon provides a base AMI which we
have built upon, using [Packer](https://github.com/hashicorp/packer) and
[Ansible](https://github.com/ansible/ansible), to create our own Amazon
Linux-derived machine image. This, and some initialization scripts, give all our
running ECS tasks access to things that all Deliveroo’s services will need, such
as software (like the [Datadog agent](https://docs.datadoghq.com/agent/) which
sends metrics to Datadog, and
[AWS Inspector](https://aws.amazon.com/inspector/), AWS's automated security
assessment service), roles, security policies, and environment variables that
we need to apply to the containers.

The process of rolling out a new machine image when an update is available, or
when we make changes to our custom machine image, is not as straightforward as

> I’m used to (I have a new-found appreciation for release management software).

Reviewer 2: Do you mean `as I would have expected` here or actually `as I'm used to` ? Just wondering if it's actually simpler with other providers?
